### PR TITLE
Lighter relative popup

### PR DESCRIPTION
**Pull request by buoyad**
This PR adds two common adapters: `FloatingBox` and `FloatingMenu`. They both aim to solve the common use case of a component that we want positioned relative to the parent on desktop, and positioned absolutely at the bottom of the screen on mobile. 

`FloatingMenu` is a simple component that uses the `FloatingBox` to position a dropdown-menu (a-la `RetentionDropdown`) in this way. `FloatingBox` is the meat of this PR, and does the following:

- On desktop, it 'wraps' the `ModalPositionRelative` HOC
  - The implementer can pass in a ref to a `React.Component` and `FloatingBox` will figure out the `ClientRect` to pass to the relative popup
  - It also takes a `Position` which it passes through to the HOC
- On mobile, it uses [react-gateway](https://github.com/cloudflare/react-gateway) to create a 'portal' as a sibling to the route tree, allowing children to fill the entire screen
  - A `GatewayDest` is added [next to the route tree](https://github.com/keybase/client/compare/danny/DESKTOP-6550-lighter-relative-popup?expand=1#diff-89fb7b09c9075f77407cf085a38c5ae3R71) on mobile to do this
  - It can be accessed globally, so there's some concern that multiple things could try to render inside it simultaneously. I put a TODO to add some level of scoping to the `GatewayDest`, and for now `FloatingBox` fills the entire screen on mobile. Additionally, `FloatingBox` is intended to be somewhat low-level, and using something like `FloatingMenu` is restrictive enough to avoid that problem.

Neither of these are meant to be directly routeable. This limits sprawl because route definitions & routing callbacks don't need to be updated/added. `FloatingMenu` is instantiated inline, and takes a `visible` prop to choose whether or not to render. It also means that the can be as tightly coupled to it's parent as needed (e.g. it can call `setState` on the parent), since they'll be mounted and unmounted together.

I have a little more testing / cleanup to do until this is ready.

- [x] ~working on mobile with route-local gateway~
  - Spent some time on this. The tab bar isn't rendered by the route tree, so the place to put the `GatewayDest` would be in nav.native.js so it can be a sibling to the whole screen (and cover the tab bar). But then it's just a global gateway with a dynamic name.

### CODE REVIEW

File: shared/common-adapters/relative-popup-hoc.desktop.js [hunk 1]
```diff
@@ -245,11 +245,21 @@ function ModalPositionRelative<PP>(
     }
 
     render() {
+      // React will complain if WrappedComponent is a HTMLElement and we try to attach these props
```
Comment (buoyad) [inline shared/common-adapters/relative-popup-hoc.desktop.js]: In `FloatingBox`, `WrappedComponent` is a div, so react complains about these unknown attributes.

**Comment by buoyad** [comment]
@keybase/react-hackers  this is ready!

File: shared/app/main-shared.native.js [hunk 2]
```diff
@@ -60,11 +60,13 @@ class Main extends Component<any> {
     }
 
     return (
-      <RenderRoute
-        routeDef={this.props.routeDef}
-        routeState={this.props.routeState}
-        setRouteState={this.props.setRouteState}
-      />
+      <Fragment>
```
Comment (chrisnojima) [inline shared/app/main-shared.native.js]: needed?

File: shared/common-adapters/relative-popup-hoc.desktop.js [hunk 3]
```diff
@@ -245,11 +245,21 @@ function ModalPositionRelative<PP>(
...
+      const noPassProps = ['targetRect', 'position', 'onClosePopup']
+      // $ForceType thinks {} is invalid for PP, we're filtering out the HOC's props
```
Comment (chrisnojima) [inline shared/common-adapters/relative-popup-hoc.desktop.js]: you can likely fix this by doing the noPassProps check in reduce and not set res[k] in that instance

File: shared/desktop/renderer/container.js [hunk 4]
```diff
@@ -1,13 +1,16 @@
...
 import {Provider} from 'react-redux'
+import {GatewayProvider} from 'react-gateway'
 import {GlobalEscapeHandler} from '../../util/escape-handler'
 
 import '../renderer/style.css'
 
 const Root = ({store, children}: any) => (
   <GlobalEscapeHandler>
-    <Provider store={store}>{children}</Provider>
+    <GatewayProvider>
```
Comment (chrisnojima) [inline shared/desktop/renderer/container.js]: do we need this on desktop?

**Comment by chrisnojima** [comment]
some small cleanup. can you take a look at the react tree and make sure its not doing any thrashing we don't expect also tnx

File: shared/app/main-shared.native.js (hunk 2 above)
Comment (buoyad) [reply shared/app/main-shared.native.js]: It is now that it's not being scoped to the route, the `GatewayDest` is a sibling to `RenderRoute`

**Comment by buoyad** [comment]
@chrisnojima reverted the route scoping and got rid of the filter (but not the flowissue :/), PTAL. I'll take a look at the tree perf now

**Comment by buoyad** [comment]
I tested on the retention picker. On desktop it only seems to update the retention picker when it's rendered (it's state is being updated). I still can't get the inspector up on mobile but the react devtools hierarchy view shows the same thing as desktop and the perf monitor doesn't flinch when I toggle it a bunch

**Pull request merged**